app/grandchallenge/evaluation/admin.py

```diff
@@ -51,6 +61,60 @@ def __init__(self, *args, **kwargs):
             ) in self.instance.read_only_fields_for_dependent_phases:
                 self.fields[field_name].disabled = True
 
+    def clean_inputs(self):
+        inputs = self.cleaned_data["inputs"]
+
+        if any(
+            elem in NON_EVALUATION_SOCKET_SLUGS
+            for elem in inputs.values_list("slug", flat=True)
+        ):
+            raise ValidationError(
+                f'Evaluation inputs cannot be of the following types: {", ".join(NON_EVALUATION_SOCKET_SLUGS)}'
+            )
+
+        return inputs
+
+    def clean(self):
+        cleaned_data = super().clean()
+
+        if self.instance.submission_kind == SubmissionKindChoices.ALGORITHM:
+            input_slugs = (
+                set(cleaned_data["inputs"].values_list("slug", flat=True))
+                if "inputs" in cleaned_data
+                else set()
+            )
+            output_slugs = (
+                set(cleaned_data["outputs"].values_list("slug", flat=True))
+                if "outputs" in cleaned_data
+                else set()
+            )
+
+            evaluation_socket_slugs = input_slugs | output_slugs
+
+            algorithm_socket_slugs = set(
+                self.instance.algorithm_interfaces.values_list(
+                    "inputs__slug", flat=True
+                )
+            ) | set(
+                self.instance.algorithm_interfaces.values_list(
+                    "outputs__slug", flat=True
+                )
+            )
```

Comment:
Can the algorithm interfaces be changed by this form? If so, I don't think that this lookup will be in sync.

Reply:
No, they cannot be changed by this form.